### DISCUSSION TITLE
Resolution of NavTargets

### DISCUSSION THREAD

**Discussion by sergejsha** [bumble-tech/appyx#219] [Q&A] [answered]
I wonder how NavTarget resolution for a deep child is done? 

For example the app shows Onboarding/O3 node (see the screenshot below). Now we want to navigate straight into Main/Messages/List node. I assume, that the the Main node is not loaded yet. At the same time the resolution should be done inside Messages node which is either not loaded. How does the library resolve the "Messages.List" NavTarget then?

<img width="864" alt="Screenshot 2022-10-15 at 16 34 21" src="https://user-images.githubusercontent.com/768721/195992099-45dfc66c-b012-4adc-8ea6-d89fba8d20a0.png">

The screenshot is taken from [Model-Driven navigation with Jetpack Compose](https://www.droidcon.com/2022/09/29/model-driven-navigation-with-jetpack-compose-from-zero-to-hero/) talk.

**Comment by zsoltk** [comment, marked as answer]
Hey @sergejsha!

For most cases, navigation can be implicit:

1. `O3` fires an event to tell `Onboarding` that it's done. `O3` can do this either either via a callback or by the `finish()` method every `Node` has.
2. `Onboarding` realises the last child has finished (see the `onChildFinished()` method you can override if you go with `finish()` instead of a callback). In its local business logic it would realise there's no next child to switch to, so it repeats the same pattern, and finishes. 
3. `Logged in` catches that `Onboarding` finished, and switches its current `NavTarget` to be `Main`
4. `Main` when constructed will load its default `NavTarget`, which can be `Messages` in your case
5. `Messages` when constructed will load its default `NavTarget`, which can be `List` in your case

This approach is usually preferred as navigation evolves as a sequence of local concerns (each `Node` in the chain just reacts to what it needs to do without knowing too much of the bigger context).

Your other approach can be workflows, by which you can chain together methods of the local api of individual `Nodes` to build a deep link resolution. This is a more explicit way of where you want to go, and can introduce unwanted coupling, especially if you use a multi-module approach. We have workflows api in the codebase but we need to catch up with the documentation for it. Meanwhile you could take a look at https://github.com/badoo/RIBs/blob/master/documentation/tree-structure-101/deep-link-workflows.md where we documented it, and the equivalent of it in Appyx should be analogous to that.
